From 1103335f8d13f29b694a8ccbec1ca5c8396ff146 Mon Sep 17 00:00:00 2001
From: Matthew Feickert <matthew.feickert@cern.ch>
Date: Sun, 1 Mar 2026 09:50:13 +0100
Subject: [PATCH] fix: Apply Henning Bahl compliation patches

* On 2026-02-27 Henning Bahl provided this series of patches for FeynHiggs
  v2.19.0 which aim to fix compilation with modern Fortran.
---
 src/EFT/Resum4H1.F       |  5 +++--
 src/EFT/beta4H.F         | 12 ++++++------
 src/EFT/beta4H4c.F       | 13 +++++++------
 src/EFT/beta4H4r.F       | 24 ++++++++++++------------
 src/Util/ODESolve.F      | 19 +++----------------
 src/Util/ODESolveRK.F    | 16 ++++++++--------
 src/Util/ODESolveStiff.F |  8 ++++----
 7 files changed, 43 insertions(+), 54 deletions(-)

diff --git a/src/EFT/Resum4H1.F b/src/EFT/Resum4H1.F
index 7ca5ab2..c44b56b 100644
--- a/src/EFT/Resum4H1.F
+++ b/src/EFT/Resum4H1.F
@@ -110,9 +110,10 @@
 
 ************************************************************************
 
-	subroutine rge(dydx, y, x)
+	subroutine rge(n, x, y, dydx)
 	implicit none
-	RealType dydx(*), y(*), x
+	integer n
+	RealType x, y(n), dydx(n)
 
 #include "const.h"
 
diff --git a/src/EFT/beta4H.F b/src/EFT/beta4H.F
index 538c45e..de2931d 100644
--- a/src/EFT/beta4H.F
+++ b/src/EFT/beta4H.F
@@ -8,10 +8,10 @@
 #include "debug.h"
 
 
-	subroutine betaSM2L(dy, y, t, userarg)
+	subroutine betaSM2L(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*), t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
 
 #include "Resum4H.h"
 
@@ -158,10 +158,10 @@
 
 ************************************************************************
 
-	subroutine betaSplit2L(dy, y, t, userarg)
+	subroutine betaSplit2L(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*), t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
 
 #include "Resum4H.h"
 
diff --git a/src/EFT/beta4H4c.F b/src/EFT/beta4H4c.F
index dac17f8..16c755c 100644
--- a/src/EFT/beta4H4c.F
+++ b/src/EFT/beta4H4c.F
@@ -1,10 +1,11 @@
 #include "debug.h"
 #include "types.h"
 
-	subroutine betaSM2Ld(dy, y, t, userarg)
+	subroutine betaSM2Ld(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*), t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
+
 #include "Resum4H.h"
 
 	RealType yg1SM2, yg2SM2, yg3SM2, yytSM2, ytbSM2, ylSM2
@@ -103,10 +104,10 @@
 	end
 
 
-	subroutine betaTHDM2Ld(dy, y, t, userarg)
+	subroutine betaTHDM2Ld(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*),t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
 
 #include "Resum4H.h"
 
diff --git a/src/EFT/beta4H4r.F b/src/EFT/beta4H4r.F
index 4e43078..1e5336d 100644
--- a/src/EFT/beta4H4r.F
+++ b/src/EFT/beta4H4r.F
@@ -6,10 +6,10 @@
 #include "types.h"
 #include "debug.h"
 
-	subroutine betaSM2Lc(dy, y, t, userarg)
+	subroutine betaSM2Lc(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*), t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
 
 #include "Resum4H.h"
 
@@ -140,10 +140,10 @@
 
 ************************************************************************
 
-	subroutine betaSplit2Lc(dy, y, t, userarg)
+	subroutine betaSplit2Lc(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*), t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
 
 #include "Resum4H.h"
 
@@ -425,10 +425,10 @@
 
 ************************************************************************
 
-	subroutine betaTHDM2L(dy, y, t, userarg)
+	subroutine betaTHDM2L(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*), t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
 
 #include "Resum4H.h"
 
@@ -1694,10 +1694,10 @@
 
 ************************************************************************
 
-	subroutine betaTHDMsplit2L(dy, y, t, userarg)
+	subroutine betaTHDMsplit2L(n, t, y, dy)
 	implicit none
-	RealType dy(*), y(*), t
-	external userarg
+	integer n
+	RealType t, y(n), dy(n)
 
 #include "Resum4H.h"
 
diff --git a/src/Util/ODESolve.F b/src/Util/ODESolve.F
index b17c7a8..7de72d9 100644
--- a/src/Util/ODESolve.F
+++ b/src/Util/ODESolve.F
@@ -1,7 +1,7 @@
 * ODESolve.F
 * ODE solver a la Numerical Recipes
 * this file is part of FeynHiggs
-* last modified 27 Feb 18 th
+* last modified 24 Feb 25 th
 
 #include "externals.h"
 #include "types.h"
@@ -15,28 +15,15 @@ c#define VERBOSE
 	implicit none
 	integer nvar
 	external f, userarg
-	RealType ystart(*), x1, x2, eps, h1
+	RealType ystart(nvar), x1, x2, eps, h1
 
 #include "FH.h"
 
 	if( stiffodesolver .eq. 1 ) then
 	  call ODESolveStiff(nvar, f, ystart, x1, x2, eps, h1, userarg)
 	elseif( stiffodesolver .eq. 2 ) then
-	  call ODESolveODEPACK(nvar, fodepack,
-     &                         ystart, x1, x2, eps, h1,userarg)
+	  call ODESolveODEPACK(nvar, f, ystart, x1, x2, eps, h1,userarg)
 	else
 	  call ODESolveRK(nvar, f, ystart, x1, x2, eps, h1, userarg)
 	endif
-
-
-	contains
-	  subroutine fodepack(neq, t, y, dy)
-	  implicit none
-	  RealType dy(*), y(*), t
-	  integer neq
-
-	  call f(dy, y, t, 0)
-
-	  end
-
 	end
diff --git a/src/Util/ODESolveRK.F b/src/Util/ODESolveRK.F
index 30d3642..3a082d8 100644
--- a/src/Util/ODESolveRK.F
+++ b/src/Util/ODESolveRK.F
@@ -1,7 +1,7 @@
 * ODESolve.F
 * ODE solver a la Numerical Recipes
 * this file is part of FeynHiggs
-* last modified 27 Feb 18 th
+* last modified 24 Feb 25 th
 
 #include "externals.h"
 #include "types.h"
@@ -15,7 +15,7 @@ c#define WARNINGS
 	implicit none
 	integer nvar
 	external f, userarg
-	RealType ystart(*), x1, x2, eps, h1
+	RealType ystart(nvar), x1, x2, eps, h1
 
 #include "FH.h"
 #define __SUBROUTINE__ "ODESolveRK"
@@ -64,7 +64,7 @@ c#define WARNINGS
 	  DPARA "x0    =", x0		ENDL
 	  DPARA "y0(1) =", y0(1)	ENDL
 #endif
-	  call f(dydx(1,1), y0, x0, userarg)
+	  call f(nvar, x0, y0, dydx(1,1), userarg)
 	  h0 = h
 	  if( (x0 + h - x2)*(x0 + h - x1) .gt. 0 ) h = x2 - x0
 
@@ -76,27 +76,27 @@ c#define WARNINGS
 * determine Cash-Karp step
 	  y(1:nvar) = y0(1:nvar) + h*
      &      B21*dydx(1:nvar,1)
-	  call f(dydx(1,2), y, x0 + A2*h, userarg)
+	  call f(nvar, x0 + A2*h, y, dydx(1,2), userarg)
 
 	  y(1:nvar) = y0(1:nvar) + h*(
      &      B31*dydx(1:nvar,1) + B32*dydx(1:nvar,2) )
-	  call f(dydx(1,3), y, x0 + A3*h, userarg)
+	  call f(nvar, x0 + A3*h, y, dydx(1,3), userarg)
 
 	  y(1:nvar) = y0(1:nvar) + h*(
      &      B41*dydx(1:nvar,1) + B42*dydx(1:nvar,2) +
      &      B43*dydx(1:nvar,3) )
-	  call f(dydx(1,4), y, x0 + A4*h, userarg)
+	  call f(nvar, x0 + A4*h, y, dydx(1,4), userarg)
 
 	  y(1:nvar) = y0(1:nvar) + h*(
      &      B51*dydx(1:nvar,1) + B52*dydx(1:nvar,2) +
      &      B53*dydx(1:nvar,3) + B54*dydx(1:nvar,4) )
-	  call f(dydx(1,5), y, x0 + A5*h, userarg)
+	  call f(nvar, x0 + A5*h, y, dydx(1,5), userarg)
 
 	  y(1:nvar) = y0(1:nvar) + h*(
      &      B61*dydx(1:nvar,1) + B62*dydx(1:nvar,2) +
      &      B63*dydx(1:nvar,3) + B64*dydx(1:nvar,4) +
      &      B65*dydx(1:nvar,5) )
-	  call f(dydx(1,6), y, x0 + A6*h, userarg)
+	  call f(nvar, x0 + A6*h, y, dydx(1,6), userarg)
 
 	  errmax = maxval(
      &      abs(h*(DC1*dydx(1:nvar,1) + DC3*dydx(1:nvar,3) +
diff --git a/src/Util/ODESolveStiff.F b/src/Util/ODESolveStiff.F
index 79948d4..f815f7c 100644
--- a/src/Util/ODESolveStiff.F
+++ b/src/Util/ODESolveStiff.F
@@ -56,7 +56,7 @@ c#define VERBOSE
 
 	  h0 = hnext
 
-	  call f(dydx(1,1), y0, x0, userarg)
+	  call f(nvar, x0, y0, dydx(1,1), userarg)
 	  if( (x0 + h0 - x2)*(x0 + h0 - x1) .gt. 0 ) h0 = x2 - x0
 	  do i = 1, nvar
 	    yscal(i) = max(1D0, abs(y0(i)))
@@ -93,7 +93,7 @@ c#define VERBOSE
 	    call LUSolveIndx(nvar, A,nvarmax, indx, g2)
 	    y0 = ysav + A31*g1 + A32*g2
 	    x0 = xsav + A3X*h0
-	    call f(dydx(1,3), y0, x0, userarg)
+	    call f(nvar, x0, y0, dydx(1,3), userarg)
 	    g3 = dydx(:,3) + h0*C3X*dfdx
      &           + (C31*g1 + C32*g2)/h0
 	    call LUSolveIndx(nvar, A,nvarmax, indx, g3)
@@ -163,11 +163,11 @@ c#define VERBOSE
 
 	dfdy(:,:) = 0D0
 
-	call fun(fvv, y, x, userarg)
+	call fun(nvar, x, y, fvv, userarg)
 	do j = 1, nvar
 	  yh = y
 	  yh(j) = yh(j) + deriv_eps
-	  call fun(fvvh, yh, x , userarg)
+	  call fun(nvar, x, yh, fvvh, userarg)
 	  do i = 1, nvar
 	    dfdy(i,j) = (fvvh(i) - fvv(i))/deriv_eps
 	  enddo
-- 
2.53.0

